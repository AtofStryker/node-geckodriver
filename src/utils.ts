import os from 'node:os'
import util from 'node:util'
import fs from 'node:fs/promises'

import decamelize from 'decamelize'

import { GECKODRIVER_DOWNLOAD_PATH } from './constants.js'
import type { GeckodriverParameters } from 'types.js'

export async function hasAccess(filePath: string) {
  return fs.access(filePath).then(() => true, () => false)
}

export function getDownloadUrl (version: string) {
  const platformIdentifier = os.platform() === 'win32'
    ? 'win'
    : os.platform() === 'darwin'
      ? 'macos'
      : 'linux'
  const arch = os.arch() === 'arm64'
    ? '-aarch64'
    : os.arch() === 'x64'
      ? '64'
      : '32'
  const ext = os.platform() === 'win32' ? '.zip' : '.tar.gz'
  return util.format(GECKODRIVER_DOWNLOAD_PATH, version, version, platformIdentifier, arch, ext)
}

const EXCLUDED_PARAMS = ['version', 'help']
export function parseParams (params: GeckodriverParameters) {
  return Object.entries(params)
    .filter(([key,]) => !EXCLUDED_PARAMS.includes(key))
    .map(([key, val]) => {
      if (typeof val === 'boolean' && !val) {
        return ''
      }
      const vals = Array.isArray(val) ? val : [val]
      return vals.map((v) => `--${decamelize(key, { separator: '-' })}${typeof v === 'boolean' ? '' : `=${v}`}`)
    })
    .flat()
    .filter(Boolean)
}
