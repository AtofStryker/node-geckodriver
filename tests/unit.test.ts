import os from 'node:os'
import { vi, test, expect } from 'vitest'

import { getDownloadUrl, parseParams } from '../src/utils.js'

vi.mock('node:os', () => ({
  default: {
    arch: vi.fn(),
    platform: vi.fn()
  }
}))

test('getDownloadUrl', () => {
  vi.mocked(os.arch).mockReturnValue('arm')
  vi.mocked(os.platform).mockReturnValue('linux')
  expect(getDownloadUrl('0.33.0')).toMatchSnapshot()
  vi.mocked(os.arch).mockReturnValue('arm64')
  vi.mocked(os.platform).mockReturnValue('linux')
  expect(getDownloadUrl('0.33.0')).toMatchSnapshot()
  vi.mocked(os.arch).mockReturnValue('arm')
  vi.mocked(os.platform).mockReturnValue('win32')
  expect(getDownloadUrl('0.33.0')).toMatchSnapshot()
  vi.mocked(os.arch).mockReturnValue('arm64')
  vi.mocked(os.platform).mockReturnValue('win32')
  expect(getDownloadUrl('0.33.0')).toMatchSnapshot()
  vi.mocked(os.arch).mockReturnValue('x64')
  vi.mocked(os.platform).mockReturnValue('darwin')
  expect(getDownloadUrl('0.33.0')).toMatchSnapshot()
  vi.mocked(os.arch).mockReturnValue('arm64')
  vi.mocked(os.platform).mockReturnValue('darwin')
  expect(getDownloadUrl('0.33.0')).toMatchSnapshot()
})

test('parseParams', () => {
  expect(parseParams({ baseUrl: 'foobar', silent: true, verbose: false, allowedIps: ['123', '321'] }))
    .toMatchSnapshot()
})
