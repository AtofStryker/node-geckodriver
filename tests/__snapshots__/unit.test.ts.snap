// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getDownloadUrl 1`] = `"https://github.com/mozilla/geckodriver/releases/download/v0.33.0/geckodriver-v0.33.0-linux32.tar.gz"`;

exports[`getDownloadUrl 2`] = `"https://github.com/mozilla/geckodriver/releases/download/v0.33.0/geckodriver-v0.33.0-linux-aarch64.tar.gz"`;

exports[`getDownloadUrl 3`] = `"https://github.com/mozilla/geckodriver/releases/download/v0.33.0/geckodriver-v0.33.0-win32.zip"`;

exports[`getDownloadUrl 4`] = `"https://github.com/mozilla/geckodriver/releases/download/v0.33.0/geckodriver-v0.33.0-win-aarch64.zip"`;

exports[`getDownloadUrl 5`] = `"https://github.com/mozilla/geckodriver/releases/download/v0.33.0/geckodriver-v0.33.0-macos64.tar.gz"`;

exports[`getDownloadUrl 6`] = `"https://github.com/mozilla/geckodriver/releases/download/v0.33.0/geckodriver-v0.33.0-macos-aarch64.tar.gz"`;

exports[`parseParams 1`] = `
[
  "--base-url=foobar",
  "--silent",
  "--allowed-ips=123",
  "--allowed-ips=321",
]
`;
